int32 x_1
int32 y_1
int32 x_2
int32 y_2